src/main/scala/chisel3/simulator/PeekPokeAPI.scala

```diff
@@ -3,11 +3,18 @@ package chisel3.simulator
 import svsim._
 import chisel3._
 
+import chisel3.experimental.{SourceInfo, SourceLine}
+import chisel3.internal.ExceptionHelpers
+
 object PeekPokeAPI extends PeekPokeAPI
 
 trait PeekPokeAPI {
-  case class FailedExpectationException[T](observed: T, expected: T, message: String)
-      extends Exception(s"Failed Expectation: Observed value '$observed' != $expected. $message")
+  case class FailedExpectationException[T](sourceInfo: SourceInfo, extraContext: Seq[String], observed: T, expected: T, message: String)
+      extends Exception(
+        s"Failed Expectation: Observed value '$observed' != $expected. " +
+        s"$message ${sourceInfo.makeMessage(x => x)}" +
+        (if (extraContext.nonEmpty) s"\n${extraContext.mkString("\n")}" else "")
+      )
```

Comment:
```suggestion
  case class FailedExpectationException[T](observed: T, expected: T, message: String)
      extends Exception(s"Failed Expectation: Observed value '$observed' != $expected. $message")
  object FailedExpectationException {
    def apply[T](observed: T, expected: T, message: String, sourceInfo: SourceInfo, extraContext: Seq[String]): FailedExpectationException[T] = {
      val fullMessage =. s"$message ${sourceInfo.makeMessage(x => x)}" +
        (if (extraContext.nonEmpty) s"\n${extraContext.mkString("\n")}" else "")
      new FailedExpectationException(observed, expected, fullMessage)
    }
  }
```
This is minor, but this is a public API and while I don't particularly care for it being public, we should try to maintain it if possible and it's easy enough to do in this case by just adding a factory method. Note I also arbitrarily changed the argument order to an order that makes more sense to me 🤷‍♀️

Reply:
This is much cleaner, thanks!

And yeah — I emulated the IR nodes having sourceInfo first, but for an exception like this it probably makes more sense having the expectations first and context after!